core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java

```diff
@@ -80,106 +86,116 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
 
         if (((DecimalType) inputType).isShort()) {
-            inputFunction = SHORT_DECIMAL_INPUT_FUNCTION;
+            inputFunction = normalizeInputMethod(SHORT_DECIMAL_INPUT_FUNCTION, boundSignature, STATE, STATE, INPUT_CHANNEL);
         }
         else {
-            inputFunction = LONG_DECIMAL_INPUT_FUNCTION;
+            inputFunction = normalizeInputMethod(LONG_DECIMAL_INPUT_FUNCTION, boundSignature, STATE, STATE, BLOCK_INPUT_CHANNEL, BLOCK_INDEX);
         }
 
         return new AggregationMetadata(
                 inputFunction,
                 Optional.empty(),
                 Optional.of(COMBINE_FUNCTION),
                 LONG_DECIMAL_OUTPUT_FUNCTION,
-                ImmutableList.of(new AccumulatorStateDescriptor<>(
-                        stateInterface,
-                        stateSerializer,
-                        new LongDecimalWithOverflowStateFactory())));
+                ImmutableList.of(
+                        new AccumulatorStateDescriptor<>(
+                        longLongStateInterface,
+                        longLongStateSerializer,
+                        longLongStateFactory),
+                        new AccumulatorStateDescriptor<>(
+                                longStateInterface,
+                                longStateSerializer,
+                                longStateFactory)));
     }
 
-    public static void inputShortDecimal(LongDecimalWithOverflowState state, Block block, int position)
+    public static void inputShortDecimal(LongLongState state, LongState overflowState, long value)
     {
-        state.setNotNull();
-
-        long[] decimal = state.getDecimalArray();
-        int offset = state.getDecimalArrayOffset();
+        state.setFirstNull(false);
 
-        long rightLow = block.getLong(position, 0);
-        long rightHigh = rightLow >> 63;
+        long[] decimal = new long[2];
+        long highValue = value >> 63;
 
         long overflow = addWithOverflow(
-                decimal[offset],
-                decimal[offset + 1],
-                rightHigh,
-                rightLow,
+                state.getFirst(),
+                state.getSecond(),
+                highValue,
+                value,
                 decimal,
-                offset);
-        state.setOverflow(Math.addExact(overflow, state.getOverflow()));
+                0);
+
+        state.setFirst(decimal[0]);
+        state.setSecond(decimal[1]);
+
+        overflowState.setValue(Math.addExact(overflow, overflowState.getValue()));
     }
 
-    public static void inputLongDecimal(LongDecimalWithOverflowState state, Block block, int position)
+    public static void inputLongDecimal(LongLongState state, LongState overflowState, Block block, int position)
     {
-        state.setNotNull();
+        state.setFirstNull(false);
 
-        long[] decimal = state.getDecimalArray();
-        int offset = state.getDecimalArrayOffset();
+        long[] decimal = new long[2];
```

Comment:
this allocates per row potentially. Could this use `Int128State` and reuse the array?